pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java

```diff
@@ -60,8 +60,15 @@ public Object handle(Request request, Response response, ActivityInstanceCreatio
             long participantId = found.getUser().getId();
 
             var activityInstanceDao = handle.attach(ActivityInstanceDao.class);
+
+            // Insert or update row for which this transaction will have to hold a lock before proceeding.
+            // This enforces that for a given activity and participant only one transaction is allowed to execute
+            // this code at a time.
+            // Lessens chance of deadlocks and ensures the counts of activity instances for a given activity are
+            // accurate so we can enforce max number of activity instances accurately
+            activityInstanceDao.upsertActivityInstanceCreationMutex(participantId, studyId, activityCode);
```

Comment:
Just a suggestion: to extract this code 
`activityInstanceDao.upsertActivityInstanceCreationMutex(participantId, studyId, activityCode);`
to method `preventMultipleCallsDeadlock()` (or invent a better name).
And put the comments as the method comments.
It could prevent method `handle()` to grow too big.

Reply:
Good idea!
b075079b4